File: terre10.js
// 24 to 12
const args = process.argv.slice(2)

if (args.length != 1) {
    console.log("Usage: terre10.js <time>")
    process.exit(1)
}

const arg = args[0]
if (arg.length != 5) {
    console.log("invalid format must be: hh:mm")
    process.exit(1)
}

let hours = parseInt(arg.substring(0, 2))
const separator = arg[2]
const minutes = parseInt(arg.substring(3, 5))

if (isNaN(hours) || separator != ":" || isNaN(minutes)) {
    console.log("invalid format must be: hh:mm")
    process.exit(1)
}

if (minutes < 0 || minutes > 59) {
    console.log("minutes must be between 0 and 59")
    process.exit(1)
}


if (hours < 0 || hours > 23) {
    console.log("hours must be between 00:00 and 23:59")
    process.exit(1)
}

let period;
if (hours == 12) {
    period = "PM"
} else if (hours > 12) {
    period = "PM"
    hours -= 12
} else {
    period = "AM"
}

// Ajoute des zéros pour un bon affichage
if (hours < 10) {
    if (minutes < 10) {
        console.log(`0${hours}:0${minutes}${period}`)
    } else {
        console.log(`0${hours}:${minutes}${period}`)
    }
} else {
    if (minutes < 10) {
        console.log(`${hours}:0${minutes}${period}`)
    } else {
        console.log(`${hours}:${minutes}${period}`)
    }
}

File: terre10.exs
# 24 to 12
defmodule Exercice do
  def convert(time_str) do
    case parse_str_to_time(time_str) do
      {12, minutes} ->
        format_time(12, minutes, "PM")

      {hours, minutes} when hours > 12 ->
        format_time(hours - 12, minutes, "PM")

      {hours, minutes} ->
        format_time(hours, minutes, "AM")

      {:error, message} ->
        IO.puts(:stderr, message)
        System.halt(1)
    end
  end

  defp parse_str_to_time(time_str) do
    try do
      [hours, minutes] =
        time_str
        |> String.split(":")
        |> Enum.map(&String.to_integer/1)

      case {hours, minutes} do
        {h, m} when h in 0..23 and m in 0..59 ->
          {:ok, {h, m}}

        _ ->
          {:error, "Time values out of range"}
      end
    rescue
      _ -> {:error, "invalid format must be: hh:mm"}
    end
  end

  defp format_time(hours, minutes, period) do
    if hours > 10 do
      if minutes > 10 do
        "#{hours}:#{minutes}#{period}"
      else
        "#{hours}:0#{minutes}#{period}"
      end
    else
      if minutes > 10 do
        "0#{hours}:#{minutes}#{period}"
      else
        "0#{hours}:0#{minutes}#{period}"
      end
    end
  end
end

case System.argv() do
  [time_str] ->
    Exercice.convert(time_str) |> IO.puts()

  _ ->
    IO.puts(:stderr, "Usage: elixir terre10.exs <date>")
    System.halt(1)
end
